libraries/HTTPClient/src/HTTPClient.cpp

```diff
@@ -0,0 +1,1119 @@
+/**
+ * ESP8266HTTPClient.cpp
+ *
+ * Created on: 02.11.2015
+ *
+ * Copyright (c) 2015 Markus Sattler. All rights reserved.
+ * This file is part of the ESP8266HTTPClient for Arduino.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ */
+
+#include <Arduino.h>
+#include <WiFi.h>
+#include <WiFiClientSecure.h>
+#include <StreamString.h>
+#include <base64.h>
+
+#include "HTTPClient.h"
+
+class TransportTraits
+{
+public:
+    virtual ~TransportTraits()
+    {
+    }
+
+    virtual std::unique_ptr<WiFiClient> create()
+    {
+        return std::unique_ptr<WiFiClient>(new WiFiClient());
+    }
+
+    virtual bool verify(WiFiClient& client, const char* host)
+    {
+        return true;
+    }
+};
+
+class TLSTraits : public TransportTraits
+{
+public:
+    TLSTraits(const String& fingerprint) :
+        _fingerprint(fingerprint)
+    {
+    }
+
+    std::unique_ptr<WiFiClient> create() override
+    {
+        // return std::unique_ptr<WiFiClient>(new WiFiClientSecure());
+    }
+
+    bool verify(WiFiClient& client, const char* host) override
+    {
+        // auto wcs = static_cast<WiFiClientSecure&>(client);
+        // return wcs.verify(_fingerprint.c_str(), host);
+    }
+
+protected:
+    String _fingerprint;
+};
+
+/**
+ * constructor
+ */
+HTTPClient::HTTPClient()
+{
+}
+
+/**
+ * destructor
+ */
+HTTPClient::~HTTPClient()
+{
+    if(_tcp) {
+        _tcp->stop();
+    }
+    if(_currentHeaders) {
+        delete[] _currentHeaders;
+    }
+}
+
+void HTTPClient::clear()
+{
+    _returnCode = 0;
+    _size = -1;
+    _headers = "";
+}
+
+
+bool HTTPClient::begin(String url, String httpsFingerprint)
+{
+    _transportTraits.reset(nullptr);
+    _port = 443;
+    if (httpsFingerprint.length() == 0) {
+        return false;
+    }
+    if (!beginInternal(url, "https")) {
+        return false;
+    }
+    _transportTraits = TransportTraitsPtr(new TLSTraits(httpsFingerprint));
```

Comment:
WiFiClientSecure doesn't support SHA1 Fingerprint anymore, because it's become insecure few days ago... More info in http://www.pcworld.com/article/3173791/security/stop-using-sha1-it-s-now-completely-unsafe.html and https://shattered.io/static/shattered.pdf